src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/BigQueryJoinDataset.java

```diff
@@ -0,0 +1,180 @@
+  private final String bqDataset;
+  private final String bqTable;
+  private final String jobId;
+  private final BigQuerySQLBuilder helper;
+  private Long numRows;
+
+  private BigQueryJoinDataset(String datasetName,
+                             JoinDefinition joinDefinition,
+                             Map<String, String> stageToTableNameMap,
+                             BigQuery bigQuery,
+                             String project,
+                             String bqDataset,
+                             String bqTable,
+                             String jobId) {
+    this.datasetName = datasetName;
+    this.joinDefinition = joinDefinition;
+    this.bigQuery = bigQuery;
+    this.project = project;
+    this.bqDataset = bqDataset;
+    this.bqTable = bqTable;
+    this.jobId = jobId;
+    this.helper = new BigQuerySQLBuilder(this.project, this.bqDataset, stageToTableNameMap);
+  }
+
+  public static BigQueryJoinDataset getInstance(SQLJoinRequest joinRequest,
+                                                Map<String, String> bqTableNamesMap,
+                                                BigQuery bigQuery,
+                                                String project,
+                                                String dataset,
+                                                String runId) {
+
+    // Get new Job ID for this push operation
+    String jobId = BigQuerySQLEngineUtils.getIdentifier();
+
+    // Build new table name for this dataset
+    String table = BigQuerySQLEngineUtils.getNewTableName(runId);
+
+    BigQueryJoinDataset instance = new BigQueryJoinDataset(joinRequest.getDatasetName(),
+                                                           joinRequest.getJoinDefinition(),
+                                                           bqTableNamesMap,
+                                                           bigQuery,
+                                                           project,
+                                                           dataset,
+                                                           table,
+                                                           jobId);
+    instance.executeJoin();
+    return instance;
+  }
+
+  private void executeJoin() {
+    TableId destinationTable = TableId.of(project, bqDataset, bqTable);
+
+    String query = getQuery(joinDefinition);
+    LOG.info("Creating table `{}` using job: {} with SQL statement: {}", bqTable, jobId, query);
+
+    // Run BigQuery job with generated SQL statement, store results in a new table, and set priority to BATCH
+    QueryJobConfiguration queryConfig =
+      QueryJobConfiguration.newBuilder(query)
+        .setDestinationTable(destinationTable)
+        .setPriority(QueryJobConfiguration.Priority.BATCH)
+        .build();
+
+    // Create a job ID so that we can safely retry.
+    JobId bqJobId = JobId.of(jobId);
+    Job queryJob = bigQuery.create(JobInfo.newBuilder(queryConfig).setJobId(bqJobId).build());
```

Comment:
believe location needs to be set here, otherwise it assumes US location

Reply:
good catch, will adjust.